src/query.rs

```diff
@@ -404,38 +403,43 @@ impl Query {
     // HD Wallets
     //
 
-    pub fn get_hd_wallets(&self) -> HashMap<Fingerprint, HDWallet> {
+    pub fn get_hd_wallets(&self) -> HashMap<DescriptorChecksum, HDWallet> {
         self.indexer.read().unwrap().watcher().wallets().clone()
     }
 
-    pub fn get_hd_wallet(&self, fingerprint: Fingerprint) -> Option<HDWallet> {
+    pub fn get_hd_wallet(&self, desc_check: DescriptorChecksum) -> Option<HDWallet> {
         self.indexer
             .read()
             .unwrap()
             .watcher()
-            .get(fingerprint)
+            .get(desc_check)
             .cloned()
     }
 
     // get the ScriptInfo entry of a derived hd key, without it necessarily being indexed
-    pub fn get_hd_script_info(&self, fingerprint: Fingerprint, index: u32) -> Option<ScriptInfo> {
+    pub fn get_hd_script_info(
+        &self,
+        desc_check: DescriptorChecksum,
+        index: u32,
+    ) -> Option<ScriptInfo> {
         let indexer = self.indexer.read().unwrap();
-        let wallet = indexer.watcher().get(fingerprint)?;
-        let key = wallet.derive(index);
-        let address = wallet.to_address(&key);
+        let wallet = indexer.watcher().get(desc_check.clone());
+        let address = wallet.unwrap().derive_address(index, &self.rpc).unwrap();
         let scripthash = ScriptHash::from(&address);
-        let origin = KeyOrigin::Derived(fingerprint, index);
+        let origin = KeyOrigin::Derived(desc_check.clone(), index);
         Some(ScriptInfo::new(scripthash, address, origin))
     }
 
-    pub fn find_hd_gap(&self, fingerprint: Fingerprint) -> Option<usize> {
+    pub fn find_hd_gap(&self, desc_check: DescriptorChecksum) -> Option<usize> {
         let indexer = self.indexer.read().unwrap();
         let store = indexer.store();
-        let wallet = indexer.watcher().get(fingerprint)?;
+        let wallet = indexer.watcher().get(desc_check)?;
         let max_funded_index = wallet.max_funded_index?; // return None if this wallet has no history at all
 
         let gap = (0..=max_funded_index)
-            .map(|derivation_index| ScriptHash::from(&wallet.derive_address(derivation_index)))
+            .map(|derivation_index| {
+                ScriptHash::from(&wallet.derive_address(derivation_index, &self.rpc).unwrap())
```

Comment:
Similarly to `get_hd_script_info()`, this now also need to be a fallible method that returns `Result` due to the reliance on RPC (and `unwrap()` also needs to be changed accordingly).

Which somewhat makes me reconsider whether we should do descriptor derivation in Rust without relying on the RPC... relying on the RPC will result in a cascade of functions that now need to be fallible, which is kind of annoying. Also, for `find_hd_gap()`, this is going to require many RPC calls, which is not ideal either.

But let's wait with this for now and give this some more thought.

Reply:
I agree that doing the derivation in rust is preferable. But I think we need to wait until apoelstra/rust-miniscript#93 lands.